quil-py/quil/program/__init__.pyi

```diff
@@ -175,6 +175,121 @@ class Program:
+
+        This schedule is computed by:
+
+        * Expanding each instruction within the block using the program's calibration definitions
+        * Resolving the `FixedSchedule` of the expanded instructions
+        * Mapping calibrated instructions back to the original instructions within this block, such that the
+          block's instruction is represented as a timespan encompassing all of its expanded instructions
+
+        If `include_zero_duration_instructions` is `True`, then the schedule will include instructions with zero duration
+        (such as `FENCE`). This may cause certain instructions to appear as unexpectedly long in duration.
+
+        :param program: The program containing the calibrations to be used to schedule this block. Generally,
+            this should be the program from which the block was extracted
+        :param include_zero_duration_instructions: Whether to include instructions with zero duration in the schedule
+
+        The following example demonstrates construction of such a schedule for a simple program without explicit control
+        flow (and thus with only one basic block):
+
+        .. example-code::
+
+            .. code-block:: python
+
+                from quil.program import Program
+
+                program = Program.parse("CZ 0 1; CZ 0 2")
+
+                print(program.to_quil())
+
+                control_flow_graph = program.control_flow_graph()
+                assert control_flow_graph.has_dynamic_control_flow() == False
+
+                basic_blocks = control_flow_graph.basic_blocks()
+                assert len(basic_blocks) == 1
+
+                schedule = blocks[0].as_fixed_schedule(program, False)
+                print(f"Duration = {schedule.duration()}")
+
+                print(schedule.items())
+
+
+         To understand why `include_zero_duration_instructions` is useful, consider a program like this:
+
+         .. example-code::
+
+            .. code-block:: text
+
+                # Two-qubit frame
+                DEFFRAME 0 1 "a":
+                    ATTRIBUTE: 1
+
+                # One-qubit frame
+                DEFFRAME 1 "a":
+                    ATTRIBUTE: 1
+
+                DEFCAL A 0:
+                    PULSE 1 "a" flat(duration: 1.0)
+
+                # The FENCE "plays" immediately upon program start at time 0
+                # The `PULSE` is blocked by the `PULSE` of `A 0` on intersecting frame 1 "a", and does not play until time 1
+                DEFCAL B 0:
+                    FENCE 0
+                    PULSE 0 1 "a" flat(duration: 1.0)
+
+                A 0
+                B 0
+
+        If ``include_zero_duration_instructions`` is ``True``, ``B 0`` will be scheduled from time 0 to time 2, because it includes the time
+        of the FENCE.
+
+        If not, it will be from time 1 to time 2, considering only the non-zero duration instructions (here, ``PULSE``) within the calibration.
```

Comment:
> So it would have a duration of 1.

I don't agree with this - nothing plays in between the FENCE and its paired PULSE, but it "ends" at the same time it starts, 0, and that's how Quil scheduling interprets it as well (for instructions which depend on it). By definition it has no duration - it's just a marker between other elements on the timeline.

> I don't think the case of users seeing the "unusually long gate times" could make any useful debugging sense

I agree, this may not, but someone (can't recall if it was you) recently specifically asked to "see when their PULSEs, CAPTUREs, and FENCEs were scheduled"), and this flag _also_ affects top-level FENCEs within the pre-expansion program. If those are missing, I'll probably hear about it.

I agree that this is a coarse implementation though. As far as seeing the "member" instructions, yeah,we could (in another PR) expose the source mapping used to actually compute the gate-model schedule.

Reply:
@mhodson-rigetti I'm confused about something in this thread after re-reading your original review on this PR:

1. it was you who had said that one of your success criteria for this was specifically seeing the FENCE timings:

> For fences, the time at which the barrier applies

2. Then, you had also called out that FENCE is zero-duration:

> this is a point-in-time operation and should have zero duration

----

I think this can be resolved by, as you say, expanding calibrations and returning the source mapping so that the user can trace back "expanded instruction time" to "original unexpanded instruction" for all instructions, and then decide how to handle that information themselves. I'll save that extension for the next PR.

Reply:
True, I wasn't being very consistent. I'm imagining what I want to visualize with this data; for a FENCE, I would draw a barrier - a vertical line on a timeline - at the latest time of any frame involved in the fence. This is "the time of the fence" if you wanted to assign it a single time. However, from a frame tracking perspective, a given frame affected by a fence goes from whatever time it is at (e.g. the time the previous pulse finished) up to this barrier time. This is basically a delay, and it might be handy to have captured that explicitly. However the data structures might not support this, and giving a FENCE a duration like this only came up when I was tracking how I would explain your new example to someone- I'd be saying that the FENCE caused "b" frame to wait from 0 to 1. I will need to deconflict this viewpoint when I write the needs statement.

Reply:
I like the tweaks you made to the example and documentation- simple and effective.

Reply:
@mhodson-rigetti after some more thought on this, I'd like to omit the flag (as you suggested and as it is now) but instead turn the default behavior to include _all_ instructions (even zero-duration) in the rolled-up timespan, because:

* While both ways can be surprising, treating all instructions equally is probably less surprising
* It suits the common case and thus will be less _frequently_ surprising as well. In most cases, FENCE shouldn't lead or lag the calibration body as in the pathological example here
* It means that pre-expansion FENCEs will show up in the overall schedule
* It means that SET- and SHIFT- instructions will as well, which lets us treat this data structure as an IR of sorts
* I intend to give the user more control over the details (as in this comment thread), so they'll have the ability to decide on a per-instruction basis what gets "rolled up" soon anyway
* Finally, what swayed me: it means that this will work as you might expect for an already-expanded pulse program; every instruction in the program will get a time span. if we skip zero-duration, some will just be missing, to no benefit.

I'd leave the illustration here to help users identify this behavior when it does happen.

Does that check out with you?

Reply:
Yeah I'm totally comfortable with that. Good call.